File: tests/leetcode/x/swift/0123.swift
import Foundation

func maxProfit(_ prices: [Int]) -> Int {
    var buy1 = -1_000_000_000
    var sell1 = 0
    var buy2 = -1_000_000_000
    var sell2 = 0
    for p in prices {
        buy1 = max(buy1, -p)
        sell1 = max(sell1, buy1 + p)
        buy2 = max(buy2, sell1 - p)
        sell2 = max(sell2, buy2 + p)
    }
    return sell2
}

let data = String(data: FileHandle.standardInput.readDataToEndOfFile(), encoding: .utf8) ?? ""
let lines = data.split(separator: "\n", omittingEmptySubsequences: false).map { String($0).replacingOccurrences(of: "\r", with: "") }
if !lines.isEmpty {
    let t = Int(lines[0].trimmingCharacters(in: .whitespacesAndNewlines)) ?? 0
    var idx = 1
    var out: [String] = []
    for _ in 0..<t {
        let n = Int(lines[idx].trimmingCharacters(in: .whitespacesAndNewlines)) ?? 0
        idx += 1
        var prices: [Int] = []
        for _ in 0..<n {
            prices.append(Int(lines[idx].trimmingCharacters(in: .whitespacesAndNewlines)) ?? 0)
            idx += 1
        }
        out.append(String(maxProfit(prices)))
    }
    FileHandle.standardOutput.write(out.joined(separator: "\n").data(using: .utf8)!)
}

File: tests/leetcode/x/pascal/0123.pas
program Main;

function MaxProfit(prices: array of LongInt): LongInt;
var i, buy1, sell1, buy2, sell2, p: LongInt;
begin
  buy1 := -1000000000; sell1 := 0; buy2 := -1000000000; sell2 := 0;
  for i := 0 to High(prices) do
  begin
    p := prices[i];
    if -p > buy1 then buy1 := -p;
    if buy1 + p > sell1 then sell1 := buy1 + p;
    if sell1 - p > buy2 then buy2 := sell1 - p;
    if buy2 + p > sell2 then sell2 := buy2 + p;
  end;
  MaxProfit := sell2;
end;

var t, tc, n, i: LongInt; prices: array of LongInt;
begin
  if EOF then Halt(0);
  ReadLn(t);
  for tc := 1 to t do begin
    ReadLn(n);
    SetLength(prices, n);
    for i := 0 to n - 1 do ReadLn(prices[i]);
    Write(MaxProfit(prices));
    if tc < t then WriteLn;
  end;
end.
